<h2>Kết hợp FBref và Sofifa Players</h2>


<h5>Phần này sẽ kết hợp các cầu thủ có trong đội hình FBref với các thông tin chi tiết hơn của cầu thủ ở trong Sofifa Players</h5>

<h4>Kết hợp đội hình trong FBref với các cầu thủ trong Sofifa</h4>

In [1]:
import pandas as pd
import numpy as np

In [2]:
#fbref
fbref_matchinfos = pd.read_csv("./fbref_modified/fbref_matchinfos_modified.csv")
fbref_matchsquad = pd.read_csv("./fbref_modified/fbref_matchsquad_modified.csv")

#sofifa
sofifa_players = pd.read_csv("./sofifa_modified/sofifa_players_attr_modified.csv")

In [3]:
fbref_matchinfos.head(2)

Match_Id          League     Season    Match_Week Home_Team  \
0  db4d7668  Premier League  2018/2019  Matchweek 38   Watford   
1  d4360fcd  Premier League  2019/2020  Matchweek 38   Burnley   

                Away_Team  Match_Date  \
0         West Ham United  2019-05-12   
1  Brighton & Hove Albion  2020-07-26   

                                 Venue_Time  Attendance  \
0   Sunday May 12, 2019, 15:00 (venue time)     20067.0   
1  Sunday July 26, 2020, 16:00 (venue time)         NaN   

                          Stadium  \
0  Vicarage Road Stadium, Watford   
1              Turf Moor, Burnley   

                                           Officials  \
0  Chris Kavanagh (Referee), Dan Cook (AR1), Sian...   
1  Jonathan Moss (Referee), Eddie Smart (AR1), Ri...   

                                                Link  
0  https://fbref.com/en/matches/db4d7668/Watford-...  
1  https://fbref.com/en/matches/d4360fcd/Burnley-...

In [4]:
fbref_matchsquad.head(2)

Match_Id     Team Is_Home_Team      Player_Name  Player_Kitnum Is_Sub
0  db4d7668  Watford          Yes       Ben Foster             26     No
1  db4d7668  Watford          Yes  Gerard Deulofeu              7     No

In [5]:
sofifa_players.head(2)

Acceleration  Age  Aggression  Agility All_Positions Attacking_Work_Rate  \
0            56   19          55       41            CB                 Low   
1            60   19          53       45         RB CB              Medium   

   Balance  Ball_Control    Birthday       Body_Type  ...  Team2_Position  \
0       47            26  2004-01-08  Lean (170-185)  ...             NaN   
1       53            38  1994-02-19   Normal (185+)  ...             NaN   

   Team2_Rating  Traits  Update_Date  Value  Vision  Volleys  Wage  Weak_Foot  \
0           NaN     NaN   2023-09-22  €100K      34       26   €2K          3   
1           NaN     NaN   2013-09-20   €35K      30       22   €1K          2   

   Weight  
0      80  
1      82  

[2 rows x 70 columns]

Xem tất cả các team của FBref và Sofifa và đổi một số Team cho trùng khớp: 

In [6]:
fbref_matchsquad['Team'].unique()

array(['Watford', 'West Ham United', 'Burnley', 'Brighton & Hove Albion',
       'Newcastle United', 'Liverpool', 'Aston Villa', 'Chelsea',
       'Wolverhampton Wanderers', 'Manchester City', 'Norwich City',
       'Crystal Palace', 'Tottenham Hotspur', 'Arsenal', 'Southampton',
       'Sheffield United', 'Leicester City', 'Manchester United',
       'Everton', 'Bournemouth', 'Fulham', 'Cardiff City',
       'Huddersfield Town', 'Brentford', 'Leeds United',
       'West Bromwich Albion', 'Luton Town', 'Nottingham Forest'],
      dtype=object)

In [7]:
different_team = set(fbref_matchsquad['Team']) - set(sofifa_players['Team1'])
different_team

{'Bournemouth'}

In [8]:
sofifa_players['Team1'].unique()

array(['Crystal Palace', 'Norwich City', 'West Bromwich Albion',
       'Manchester United', 'Manchester City', 'Cardiff City',
       'Sheffield United', 'Aston Villa', 'Wolverhampton Wanderers',
       'Brighton & Hove Albion', 'Burnley', 'Fulham', 'West Ham United',
       'Leeds United', 'Newcastle United', 'Southampton', 'Chelsea',
       'Arsenal', 'Watford', 'AFC Bournemouth', 'Tottenham Hotspur',
       'Liverpool', 'Huddersfield Town', 'Leicester City', 'Stoke City',
       'Everton', 'Swansea City', 'Middlesbrough', 'Hull City',
       'Sunderland', 'Queens Park Rangers', 'Nottingham Forest',
       'Republic of Ireland', 'Austria', 'Brentford', 'Wales',
       'Australia', 'Scotland', 'Luton Town', 'Romania', 'Sweden',
       'Czech Republic', 'Norway', 'Finland', 'Germany', 'Cameroon',
       'Switzerland', 'Iceland', 'Ukraine', 'Belgium', 'United States',
       'Argentina', 'Denmark', 'Uruguay', 'Portugal', 'Hungary', 'Poland',
       'England', 'Greece', 'Korea Republic'

In [9]:
sofifa_players['Team1'] = sofifa_players['Team1'].replace('AFC Bournemouth', 'Bournemouth')
sofifa_players['Team1'].unique()

array(['Crystal Palace', 'Norwich City', 'West Bromwich Albion',
       'Manchester United', 'Manchester City', 'Cardiff City',
       'Sheffield United', 'Aston Villa', 'Wolverhampton Wanderers',
       'Brighton & Hove Albion', 'Burnley', 'Fulham', 'West Ham United',
       'Leeds United', 'Newcastle United', 'Southampton', 'Chelsea',
       'Arsenal', 'Watford', 'Bournemouth', 'Tottenham Hotspur',
       'Liverpool', 'Huddersfield Town', 'Leicester City', 'Stoke City',
       'Everton', 'Swansea City', 'Middlesbrough', 'Hull City',
       'Sunderland', 'Queens Park Rangers', 'Nottingham Forest',
       'Republic of Ireland', 'Austria', 'Brentford', 'Wales',
       'Australia', 'Scotland', 'Luton Town', 'Romania', 'Sweden',
       'Czech Republic', 'Norway', 'Finland', 'Germany', 'Cameroon',
       'Switzerland', 'Iceland', 'Ukraine', 'Belgium', 'United States',
       'Argentina', 'Denmark', 'Uruguay', 'Portugal', 'Hungary', 'Poland',
       'England', 'Greece', 'Korea Republic', 'F

FBref_MatchSquad mới cập nhật thêm ngày diễn ra trận đấu Match_Date

In [10]:
fbref_matchsquad_new = pd.merge(fbref_matchsquad, fbref_matchinfos, on='Match_Id')
fbref_matchsquad_new = fbref_matchsquad_new.loc[:,['Match_Id', 'Match_Date', 'Team',
                                                             'Player_Name', 'Player_Kitnum', 'Is_Sub']]
fbref_matchsquad_new.head(2)

Match_Id  Match_Date     Team      Player_Name  Player_Kitnum Is_Sub
0  db4d7668  2019-05-12  Watford       Ben Foster             26     No
1  db4d7668  2019-05-12  Watford  Gerard Deulofeu              7     No

Sắp xếp Sofifa_playes theo ngày cập nhật Update_Date giảm dần và tăng dần

In [11]:
sofifa_players_desc_update_date = sofifa_players.sort_values(by = 'Update_Date', ascending=False)
sofifa_players_asc_update_date = sofifa_players.sort_values(by = 'Update_Date', ascending=True)

* function() ClosestUpdateDate: trả về ngày Update_Date gần nhất cho Match_Date (ngày diễn ra match)
* function() SofifaPlayersEqualUpdateDate: trả về dataframe tất cả các sofifa player với 1 Update_Date

In [12]:
#return closest Update_Date: date-time
def ClosestUpdateDate(sofifa_players_desc_update_date, match_date):
    closest_update_date = sofifa_players_desc_update_date[sofifa_players_desc_update_date['Update_Date'] 
                                                        <= match_date]['Update_Date'].iloc[0]
    return closest_update_date

#return dataframe
def SofifaPlayersEqualUpdateDate(sofifa_players_desc_update_date, closest_update_date):
    sofia_players_one_update_version = sofifa_players_desc_update_date[sofifa_players_desc_update_date['Update_Date'] 
                                                                    == closest_update_date]
    return sofia_players_one_update_version

* function() SofifaPlayer: trả về player với id, position và overall rating ứng với mỗi match
* function() SofifaPlayer_ByName: trả về player theo name

In [13]:
#return dataframe:
def SofifaPlayer(sofifa_update_version, team_fbref, player_kitnum_fbref):
    player = sofifa_update_version[((sofifa_update_version['Team1'] == team_fbref) | 
                                            (sofifa_update_version['Team2'] == team_fbref))
                                            & 
                                            ((sofifa_update_version['Team1_Kitnum'] == player_kitnum_fbref) | 
                                                (sofifa_update_version['Team2_Kitnum'] 
                                                == player_kitnum_fbref))][['Sofifa_Id', 'All_Positions',
                                                                           'Overall_Rating', 'Update_Date']]
    return player

#return dataframe:
def SofifaPlayer_ByName(sofifa_update_version, player_name):
    player = sofifa_update_version[sofifa_update_version['Player_Name'] 
                                            == player_name][['Sofifa_Id',
                                                             'All_Positions',
                                                             'Overall_Rating', 
                                                             'Update_Date']]
    return player

    

- function() PLayerNotFound: kiểm tra trường hợp player trong match không có trong phiên bản sofifa gần nhất

In [14]:
#return player_id: integer
def PLayerNotFound(sofifa_players_asc_update_date,closest_update_date, player_name):

    player = pd.DataFrame()

    dates_last = sofifa_players_asc_update_date[sofifa_players_asc_update_date['Update_Date'] 
                                                            > closest_update_date]['Update_Date'].unique()
    if(len(dates_last)>0):
        for k in range(0,len(dates_last)):
            sofia_players_more_update_date = sofifa_players_asc_update_date[sofifa_players_asc_update_date['Update_Date'] 
                                                                        == dates_last[k]]

            player = SofifaPlayer_ByName(sofia_players_more_update_date, player_name)
            
            if not player.empty:
                break

    return player

Cập nhật thêm cho MatchSquad với sofifa_id player, vị trí, overall rating và phiên bản player

In [15]:
new_attrs = pd.DataFrame({'Sofifa_Id': [],
                          'Player_Position': [],
                          'Player_Overall_Rating': [],
                          'Player_Update_Date': []})
list_sofifa_id = []
list_player_position = []
list_player_overall_rating = []
list_update_date = []

In [16]:
# Update_Date attribute
# match_date_i = fbref_matchsquad_new.iloc[i, fbref_matchsquad_new.columns.get_loc('Match_Date')]

Xác định ứng với mỗi match, tại thời điểm Update_Date của player gần nhất, với mỗi Team - đội bóng và Player_Kitnum - số áo player trong đội ta có thể tìm được player tương ứng trong Sofifa

In [17]:
for i in range(0, len(fbref_matchsquad_new)):
    # 
    match_date_i = fbref_matchsquad_new.iloc[i, fbref_matchsquad_new.columns.get_loc('Match_Date')]

    closest_update_date_i = ClosestUpdateDate(sofifa_players_desc_update_date, match_date_i)

    # Sofifa_Id attribute
    # ** team **
    team_fbref_i = fbref_matchsquad_new.iloc[i, fbref_matchsquad_new.columns.get_loc('Team')]
    # ** player_kit_num **
    player_kitnum_fbref_i = fbref_matchsquad_new.iloc[i, fbref_matchsquad_new.columns.get_loc('Player_Kitnum')]

    sofia_players_this_update_version = SofifaPlayersEqualUpdateDate(sofifa_players_desc_update_date, closest_update_date_i)  
    
    player = SofifaPlayer(sofia_players_this_update_version, team_fbref_i, player_kitnum_fbref_i)
    
    player_id = np.nan
    player_position = np.nan
    player_overall_rating = np.nan
    player_update_date = np.nan

    #if player.empty:
        #player = PLayerNotFound(sofifa_players_asc_update_date,closest_update_date_i, team_fbref_i, player_kitnum_fbref_i)
    
    if not player.empty:
        player_id = player['Sofifa_Id'].iloc[0]
        player_position = player['All_Positions'].iloc[0]#.split(' ')[0]
        player_overall_rating = player['Overall_Rating'].iloc[0]
        player_update_date = player['Update_Date'].iloc[0]
    

    list_sofifa_id.append(player_id)

    #Player_Position attribute
    list_player_position.append(player_position)

    #Player_Overall_Rating attribute
    list_player_overall_rating.append(player_overall_rating)

    #Player_Update_Date attribute
    list_update_date.append(player_update_date)

In [18]:
new_attrs['Sofifa_Id'] = list_sofifa_id
new_attrs['Player_Position'] = list_player_position
new_attrs['Player_Overall_Rating'] = list_player_overall_rating
new_attrs['Player_Update_Date'] = list_update_date

Kết quả kết hợp giữa Fbref và Sofifa

In [19]:
matchsquad_players = pd.concat([fbref_matchsquad_new, new_attrs], axis=1)
matchsquad_players

Match_Id  Match_Date     Team           Player_Name  Player_Kitnum  \
0      db4d7668  2019-05-12  Watford            Ben Foster             26   
1      db4d7668  2019-05-12  Watford       Gerard Deulofeu              7   
2      db4d7668  2019-05-12  Watford           Troy Deeney              9   
3      db4d7668  2019-05-12  Watford        Craig Cathcart             15   
4      db4d7668  2019-05-12  Watford    Abdoulaye Doucouré             16   
...         ...         ...      ...                   ...            ...   
76289  e62f6e78  2022-08-05  Arsenal         Cédric Soares             17   
76290  e62f6e78  2022-08-05  Arsenal          Nicolas Pépé             19   
76291  e62f6e78  2022-08-05  Arsenal  Albert Sambi Lokonga             23   
76292  e62f6e78  2022-08-05  Arsenal          Reiss Nelson             24   
76293  e62f6e78  2022-08-05  Arsenal        Mohamed Elneny             25   

      Is_Sub  Sofifa_Id Player_Position  Player_Overall_Rating  \
0         No   163155.0              GK                   79.0   
1         No   193061.0          LM CAM                   80.0   
2         No   183125.0              ST                   78.0   
3         No   177578.0              CB                   74.0   
4         No   208135.0          CDM CM                   81.0   
...      ...        ...             ...                    ...   
76289    Yes   201118.0              RB                   75.0   
76290    Yes   226110.0              RM                   80.0   
76291    Yes   241928.0          CM CDM                   75.0   
76292    Yes        NaN             NaN                    NaN   
76293    Yes   211454.0          CDM CM                   76.0   

      Player_Update_Date  
0             2019-05-02  
1             2019-05-02  
2             2019-05-02  
3             2019-05-02  
4             2019-05-02  
...                  ...  
76289         2022-07-06  
76290         2022-07-06  
76291         2022-07-06  
76292                NaN  
76293         2022-07-06  

[76294 rows x 10 columns]

Xuất ra CSV

In [20]:
matchsquad_players.to_csv('./merge/matchsquad_players.csv', index=False)

Kiểm tra, bổ sung các player nAn bằng phương pháp tìm theo tên

In [38]:
#dành cho đã chạy trước
matchsquad_players = pd.read_csv("./merge/matchsquad_players.csv")
matchsquad_players.head(3)

Match_Id  Match_Date     Team      Player_Name  Player_Kitnum Is_Sub  \
0  db4d7668  2019-05-12  Watford       Ben Foster             26     No   
1  db4d7668  2019-05-12  Watford  Gerard Deulofeu              7     No   
2  db4d7668  2019-05-12  Watford      Troy Deeney              9     No   

   Sofifa_Id Player_Position  Player_Overall_Rating Player_Update_Date  
0   163155.0              GK                   79.0         2019-05-02  
1   193061.0          LM CAM                   80.0         2019-05-02  
2   183125.0              ST                   78.0         2019-05-02

* Kiểm tra số dòng bị thiếu player so với tổng các dòng

In [39]:
print("Số player NaN: ", matchsquad_players['Sofifa_Id'].isna().sum())
print("Tổng các dòng: ", len(matchsquad_players))

Số player NaN:  4422
Tổng các dòng:  76294


* Kiểm tra một số player bị trùng tên

In [40]:
unique_name_count = sofifa_players.groupby(['Player_Name', 'Team1'])['Sofifa_Id'].nunique().loc[lambda x: x > 1]
unique_name_count

Player_Name     Team1           
C. Chukwuemeka  Aston Villa         2
Gabriel         Arsenal             2
J. Maddison     Leicester City      2
J. Murphy       Norwich City        2
K. Phillips     Crystal Palace      2
S. Ameobi       Newcastle United    2
Name: Sofifa_Id, dtype: int64

* Lưu các tên bị trùng

In [41]:
same_name = unique_name_count.index.get_level_values('Player_Name').tolist()
same_name

['C. Chukwuemeka',
 'Gabriel',
 'J. Maddison',
 'J. Murphy',
 'K. Phillips',
 'S. Ameobi']

* Tạo dataframe chứa các 4422 dòng bị thiếu player

In [42]:
matchsquad_NaN_players = matchsquad_players[matchsquad_players['Sofifa_Id'].isna()]
matchsquad_NaN_players

Match_Id  Match_Date              Team               Player_Name  \
50     d4360fcd  2020-07-26           Burnley              Max Thompson   
54     d4360fcd  2020-07-26           Burnley  Anthony Driscoll-Glennon   
55     d4360fcd  2020-07-26           Burnley             Joel Mumbongo   
94     b6b1209d  2020-07-26  Newcastle United             Kelland Watts   
95     b6b1209d  2020-07-26  Newcastle United                Jack Young   
...         ...         ...               ...                       ...   
76273  e62f6e78  2022-08-05    Crystal Palace               Luke Plange   
76281  e62f6e78  2022-08-05           Arsenal        Gabriel Martinelli   
76282  e62f6e78  2022-08-05           Arsenal            William Saliba   
76287  e62f6e78  2022-08-05           Arsenal             Eddie Nketiah   
76292  e62f6e78  2022-08-05           Arsenal              Reiss Nelson   

       Player_Kitnum Is_Sub  Sofifa_Id Player_Position  Player_Overall_Rating  \
50                33    Yes        NaN             NaN                    NaN   
54                45    Yes        NaN             NaN                    NaN   
55                47    Yes        NaN             NaN                    NaN   
94                49    Yes        NaN             NaN                    NaN   
95                60    Yes        NaN             NaN                    NaN   
...              ...    ...        ...             ...                    ...   
76273             48    Yes        NaN             NaN                    NaN   
76281             11     No        NaN             NaN                    NaN   
76282             12     No        NaN             NaN                    NaN   
76287             14    Yes        NaN             NaN                    NaN   
76292             24    Yes        NaN             NaN                    NaN   

      Player_Update_Date  
50                   NaN  
54                   NaN  
55                   NaN  
94                   NaN  
95                   NaN  
...                  ...  
76273                NaN  
76281                NaN  
76282                NaN  
76287                NaN  
76292                NaN  

[4422 rows x 10 columns]

* Đổi PLayer_Name theo định dạng A. B 

In [43]:
new_player_name_column = pd.DataFrame({'Player_Name_New': []})
new_player_name_column['Player_Name_New'] = matchsquad_NaN_players['Player_Name'].str.split().str[0].str[0] + '. ' + matchsquad_NaN_players['Player_Name'].str.split().str[-1]

new_player_name_column.head(3)

Player_Name_New
50          M. Thompson
54  A. Driscoll-Glennon
55          J. Mumbongo

In [44]:
matchsquad_NaN_players = pd.concat([matchsquad_NaN_players, new_player_name_column], axis=1)
matchsquad_NaN_players.head(3)

Match_Id  Match_Date     Team               Player_Name  Player_Kitnum  \
50  d4360fcd  2020-07-26  Burnley              Max Thompson             33   
54  d4360fcd  2020-07-26  Burnley  Anthony Driscoll-Glennon             45   
55  d4360fcd  2020-07-26  Burnley             Joel Mumbongo             47   

   Is_Sub  Sofifa_Id Player_Position  Player_Overall_Rating  \
50    Yes        NaN             NaN                    NaN   
54    Yes        NaN             NaN                    NaN   
55    Yes        NaN             NaN                    NaN   

   Player_Update_Date      Player_Name_New  
50                NaN          M. Thompson  
54                NaN  A. Driscoll-Glennon  
55                NaN          J. Mumbongo

* Bổ sung các player sử dụng như bước ở phần trước với tìm theo tên player_name

In [45]:
new_attrs = pd.DataFrame({'Sofifa_Id': [],
                          'Player_Position': [],
                          'Player_Overall_Rating': [],
                          'Player_Update_Date': []})
list_sofifa_id = []
list_player_position = []
list_player_overall_rating = []
list_update_date = []

In [46]:
for i in range(0, len(matchsquad_NaN_players)):
    # 
    match_date_i = matchsquad_NaN_players.iloc[i, matchsquad_NaN_players.columns.get_loc('Match_Date')]

    closest_update_date_i = ClosestUpdateDate(sofifa_players_desc_update_date, match_date_i)

    player_name_new_fbref_i = matchsquad_NaN_players.iloc[i, matchsquad_NaN_players.columns.get_loc('Player_Name_New')]
    
    #check name nằm trong list same_name
    #if player_name_new_fbref_i in same_name:
    #    continue

    sofia_players_this_update_version = SofifaPlayersEqualUpdateDate(sofifa_players_desc_update_date, closest_update_date_i)  
    
    player = SofifaPlayer_ByName(sofia_players_this_update_version, player_name_new_fbref_i)
    
    player_id = np.nan
    player_position = np.nan
    player_overall_rating = np.nan
    player_update_date = np.nan

    if player.empty:
        player = PLayerNotFound(sofifa_players_desc_update_date,closest_update_date_i, player_name_new_fbref_i)
    
    if not player.empty:
        player_id = player['Sofifa_Id'].iloc[0]
        player_position = player['All_Positions'].iloc[0]#.split(' ')[0]
        player_overall_rating = player['Overall_Rating'].iloc[0]
        player_update_date = player['Update_Date'].iloc[0]

    list_sofifa_id.append(player_id)

    #Player_Position attribute
    list_player_position.append(player_position)

    #Player_Overall_Rating attribute
    list_player_overall_rating.append(player_overall_rating)

    #Player_Update_Date attribute
    list_update_date.append(player_update_date)

    #print(i,match_date_i, sofia_players_this_update_version['Update_Date'].iloc[0], player_name_new_fbref_i, player_id, player_update_date)

In [55]:
matchsquad_NaN_players['Sofifa_Id'] = list_sofifa_id
matchsquad_NaN_players['Player_Position'] = list_player_position
matchsquad_NaN_players['Player_Overall_Rating'] = list_player_overall_rating
matchsquad_NaN_players['Player_Update_Date'] = list_update_date

* Số lượng player NaN giảm đi đáng kể

In [56]:
len(matchsquad_NaN_players[matchsquad_NaN_players['Sofifa_Id'].isna()])

850

In [57]:
len(matchsquad_players)

76294

In [58]:
matchsquad_NaN_players = matchsquad_NaN_players.drop('Player_Name_New', axis=1)

KeyError: "['Player_Name_New'] not found in axis"

In [59]:
complete_squad_player =  matchsquad_players.fillna(matchsquad_NaN_players)
complete_squad_player

Match_Id  Match_Date     Team           Player_Name  Player_Kitnum  \
0      db4d7668  2019-05-12  Watford            Ben Foster             26   
1      db4d7668  2019-05-12  Watford       Gerard Deulofeu              7   
2      db4d7668  2019-05-12  Watford           Troy Deeney              9   
3      db4d7668  2019-05-12  Watford        Craig Cathcart             15   
4      db4d7668  2019-05-12  Watford    Abdoulaye Doucouré             16   
...         ...         ...      ...                   ...            ...   
76289  e62f6e78  2022-08-05  Arsenal         Cédric Soares             17   
76290  e62f6e78  2022-08-05  Arsenal          Nicolas Pépé             19   
76291  e62f6e78  2022-08-05  Arsenal  Albert Sambi Lokonga             23   
76292  e62f6e78  2022-08-05  Arsenal          Reiss Nelson             24   
76293  e62f6e78  2022-08-05  Arsenal        Mohamed Elneny             25   

      Is_Sub  Sofifa_Id Player_Position  Player_Overall_Rating  \
0         No   163155.0              GK                   79.0   
1         No   193061.0          LM CAM                   80.0   
2         No   183125.0              ST                   78.0   
3         No   177578.0              CB                   74.0   
4         No   208135.0          CDM CM                   81.0   
...      ...        ...             ...                    ...   
76289    Yes   201118.0              RB                   75.0   
76290    Yes   226110.0              RM                   80.0   
76291    Yes   241928.0          CM CDM                   75.0   
76292    Yes   231448.0           RW LW                   74.0   
76293    Yes   211454.0          CDM CM                   76.0   

      Player_Update_Date  
0             2019-05-02  
1             2019-05-02  
2             2019-05-02  
3             2019-05-02  
4             2019-05-02  
...                  ...  
76289         2022-07-06  
76290         2022-07-06  
76291         2022-07-06  
76292         2023-10-04  
76293         2022-07-06  

[76294 rows x 10 columns]

Xuất kết quả ra CSV

In [60]:
complete_squad_player.to_csv('./merge/matchsquad_players_completed.csv', index=False)

In [53]:
#sofifa_players[((sofifa_players['Team1'] == 'Southampton') | 
#                                        (sofifa_players['Team2'] == 'Southampton'))
#                                       & 
#                                        ((sofifa_players['Team1_Kitnum'] == 36) | 
#                                            (sofifa_players['Team2_Kitnum'] 
#                                            == 36))][['Player_Name', 'Sofifa_Id', 'All_Positions',
#                                                                        'Overall_Rating', 'Update_Date']]

In [54]:
#2020-07-26
#sofifa_players[sofifa_players['Player_Name'] == 'M. Thompson']['Update_Date'].unique()